crates/autopilot/src/arguments.rs

```diff
@@ -208,6 +209,11 @@ pub struct Arguments {
     )]
     pub solve_deadline: Duration,
 
+    /// A list of fee policies in the following format: `[LimitOrderOutOfMarket:0.1:0.2,LimitOrderInMarket::2.5]`
+    /// 
+    #[clap(long, env, use_value_delimiter = true)]
+    pub fee_policies: Vec<FeePolicy>,
```

Comment:
> What would it semantically mean to define multiple entries with the same policy but different parameters?

It enables us to take a quote deviation fee, and after that another fee as percent of volume (meaning we can chain multiple different fees for a single order). IIUC, this is something we have to support for integrators.

One could argue that we should summarize all of them to a single value and send that single value to the `driver`, but can't do that since not all types of fees are known upfront (volume percent and absolute fee are, quote deviation not).

Then, since db migrations are painful, I decided to model the db table for fee polices in a way that supports all foreseeable types of policies and not just for price improvement type. [That design ](https://github.com/cowprotocol/services/blob/d2ac47e1ad411be5d89a019e831e24f194e1af27/database/sql/V057__add_fee_policy.sql) was created to be memory efficient and decent performance wise (I could have created a separate table for each fee type but then the fetching would be slower).

Why

> pub enum FeePolicy {
>     LimitOrderOutOfMarket(FeePolicyParams),
>     LimitOrderInMarket(FeePolicyParams),
> }

Because I expect to add a third type - TWAP orders.

I hope my reasoning are clearer now. Note that this PR is not yet ready for merging, I just need to get the approval for the last [commit](https://github.com/cowprotocol/services/commit/f783d68a4dfdd528d618bc6c724fde021348065b) that my new logic is ok.

Reply:
Also, I hoped that I could support these two types of fees in the first iteration, since it will be trivial with this design:

Reply:
> IIUC, this is something we have to support for integrators.

My vision for integrators is that they specify their fee policy as part of the app data and it doesn't become part of the arguments we start the core protocol with.

The core protocol will experiment with different fees (price improvements, volume based, etc) but I foresee its fee config to be relatively concise and not a chain of many different fee compositions. Also, the protocol can only charge fees on order types that are a first class citizen in the core protocol (limit orders).

> Because I expect to add a third type - TWAP orders.

TWAP orders are not part of the core CoW Protocol. They are a set of orders that semantically belong together and are orchestrated and placed from a watch tower, but as far as CoW protocol is concerned users are simply placing limit orders in awkwardly consistent periods.

So my goal here is to split what the protocol boots up with as a fee policy from the different types of protocol fees we can attach to each order that is sent to the solvers.

E.g. the protocol may define that we want to take 0.1% volume fee on all orders. This is what the autopilot should be started with.

If an integrator (e.g. CoW Swap's TWAP feature or Metamask) wants to charge 0.85% volume fee on top, then they can create an order specifying this in the app data and what we end up sending to the driver is two fees (0.1% for the protocol, 0.85% for the integrator), both of which will be charged from solvers if they settle this order at the end. Potentially the protocol will demand a cut (cf. Apple charging 30% of all app store purchases) from the integrator fees.

To sum up, I think we should have a `FeePolicy` struct on the autopilot arguments and the a `protocol_fee: Vec<Fee>` on the order itself.

Reply:
Ok, not using configuration type for autopilot domain makes sense, will split it.

Do you think we should keep the possibility that 2 different policies (for example quote deviation and volume based) can be sent for a single order to the driver?

If you don't like having a third enum variant for TWAP orders, how would you model a situation where we set quote deviation policy for limit orders and volume based policy for TWAP orders using:

> price improvement factor
> volume % cap
> absolute cap
> whether or not to include in-market priced limit orders